### ISSUE TITLE
Steinhafels - Bedding Only Delivery Charges

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2432]
_Imported from trac ticket 2432.
Created by **dzuern**
Opened in trac: Aug 2, 2013, 2:30:21 PM
Last modified in trac: Jun 1, 2015, 4:11:36 PM_
***
_by dzuern:_
Delivery charges mainly should come from the zone.  Bedding has some exceptions.  

**Bedding Only is defined as:**
 - STOCKEDITEM = Y
 - PRODUCTCLASSID >= 930
   - Except if PRODUCTCLASSID = 990 and SUBCLASSID in ( 996, 996m, 997, 990, 990A, 990B, 990C)
 - A _bedding only order_ cannot not have sku FINLSELCT on the order.
 - Can have non-inventory items on the order but no other inventory type items.

**Calculate the subtotal:**
  - Sum of "Bedding Only Sale" = PRODUCTCLASSID >= 930 excludes PRODUCTCLASSID in (964, 968, 949, 990)

**Pricing(Cannot be negative):**
- If a “Bedding Order Only” sale and subtotal <= $498, subtract $50 from the zones del_chg 
- If a “Bedding Order Only” sale and subtotal > $498, subtract $89.95 from the charge (which is currently the lowest zone delivery charge and subject to change). 
- Allow delivery charge to be changed, but must provide a reason (ie, comment) and enter apprv pswd

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Aug 12, 2013, 3:53:00 PM:_

SDM is the zone representing Same Day Mattress.  
 - SDM can only be selected for Bedding Only orders.  
 - SDM cannot be used after 3pm for today's date.

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 16, 2013, 9:48:42 AM:_

Replying inline to [description](ticket:2432):
> Delivery charges mainly should come from the zone.  Bedding has some exceptions.  
> 
> '''Bedding Only is defined as:'''
>  - STOCKEDITEM = Y
>  - PRODUCTCLASSID >= 930



This is an alphanumeric string. Therefore, all alphabetic minor codes, like 'ABC' will be >= '930'.  If two digit minor codes are used, some of these are also >= '930', such as '94', '95', etc.


>    - Except if PRODUCTCLASSID = 990 and SUBCLASSID in ( 996, 996m, 997, 990, 990A, 990B, 990C)
>  - A ''bedding only order'' cannot not have sku FINLSELCT on the order.
>  - Can have non-inventory items on the order but no other inventory type items.


Definition of "inventory" and "non-inventory" is needed here.  type == 'INV' or stockeditem = 'Y', etc?  

> 
> '''Calculate the subtotal:'''
>   - Sum of "Bedding Only Sale" = PRODUCTCLASSID >= 930 excludes PRODUCTCLASSID in (964, 968, 949, 990)
> 
> '''Pricing(Cannot be negative):'''
> - If a “Bedding Order Only” sale and subtotal <= $498, subtract $50 from the zones del_chg 
> - If a “Bedding Order Only” sale and subtotal > $498, subtract $89.95 from the charge (which is currently the lowest zone delivery charge and subject to change). 
> - Allow delivery charge to be changed, but must provide a reason (ie, comment) and enter apprv pswd

We cannot force a comment, but we can remind manager to add comment.  The system will log which manager approved the changed.  Also, modifying anything that thereafter changes the delivery charge, including splitting the ticket, will reset the delivery change and require manager re-approval to re-modify it away from Steinhafel's calculation.  This is extremely difficult to program otherwise, but we suspect changing such fundamentals _should_, in Steinhafel's best interest, require re-approval anyway.

Are these ticket (delivery) based or entire-order based (what we term "customer order")?  In other words, if the A ticket qualifies, but the main ticket does not, what is the total delivery charge?

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 16, 2013, 12:48:28 PM:_

Replying to [kbower](comment:3):
> Are these ticket (delivery) based or entire-order based (what we term "customer order")?  In other words, if the A ticket qualifies, but the main ticket does not, what is the total delivery charge?


by CustomerOrder

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Sep 16, 2013, 4:43:41 PM:_

The question was "How do you determine an inventory vs. non-inventory?" 
In the current code, it specifically looks for INV in the itm_tp_cd field.  And NOT at the Y/N field.

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 20, 2013, 8:11:00 AM:_

Below are the 64 minor codes (productclasses) that are " >= '930' "

Note that "Except if PRODUCTCLASSID = 990 and SUBCLASSID in ( 996, 996m, 997, 990, 990A, 990B, 990C) "

looks like it would still include:

996   996   ADDITIONAL DELIVERY CHARGES
998   998   FABRIC PROTECTION
999   998   EXTENDED WARRANTY

Do you need to refine this to something else, perhaps to this?:

'930' <= productclassid < '991' Except if PRODUCTCLASSID = 990 and SUBCLASSID in ( 996, 996m, 997, 990, 990A, 990B, 990C)

```
SQL> select * from productclasses where PRODUCTCLASSID >= '930';

PRODU SUBCA DESCRIPTION
----- ----- ----------------------------------------
930   930   SEALY TWIN
931   930   SEALY TWIN FOUNDATION
932   930   SEALY BUNKIE
933   930   SEALY TWIN-XLONG
934   920   STOW-A-WAY STORAGE FDN
935   930   SEALY FULL
936   930   SEALY FULL FOUNDATION
937   930   SEALY FULL X-LONG
938   930   SEALY FULL X-LONG FOUNDATION
939   930   SEALY QUEEN
940   930   SEALY QUEEN FOUNDATION
941   930   SEALY SPLIT QUEEN FOUNDATION
942   930   SEALY KING OR CAL.KING
943   930   SEALY TWIN X-LONG FOUNDATION
944   930   STEARNS & FOSTER MATTRESS
945   930   STEARNS & FOSTER FOUNDATION
946   930   MEMORY FOAM
947   930   DREAMS POSITIONAL MATTS
948   920   FABRIC ITEMS
949   920   BEDFRAMES & FRAME ACCESS.
950   930   DREAMS TWIN
951   930   DREAMS TWIN FOUNDATION
952   930   DREAMS BUNKIE PLATFORM
953   930   DREAMS TWIN X-LONG
954   930
955   930   DREAMS FULL
956   930   DREAMS FULL FOUNDATION
957   930   DREAMS FULL X-LONG
958   930   DREAMS FULL X-LONG FOUNDATION
959   930   DREAMS QUEEN
960   930   DREAMS QUEEN FOUNDATION
961   930   DREAMS SPLIT QUEEN FOUNDATION
962   930   DREAMS KING
963   930   DREAMS TWIN X-LONG FOUNDATION
964   998   MATTRESS PROTECTORS
965   920   PILLOW PROTECTORS,SHEETS,BLANKETS
966   930   TEMPURPEDIC MATT
967   930   TEMPURPEDIC FDNS
968   920   PILLOWS,ETC.
969   920   ADJ BASE ACCESSORIES
970   930   SIMMONS TWIN
971   930   SIMMONS TWIN FOUNDATION
972   930   SIMMONS PALLET
973   930   SIMMONS TWIN X-LONG
974   930   SIMMONS
975   930   SIMMONS FULL
976   930   SIMMONS FULL FOUNDATION
977   930   ADJUSTABLE BASES
978   930   SIMMONS FULL X LONG FOUNDATION
979   930   SIMMONS QUEEN
980   930   SIMMONS QUEEN FOUNDATION
981   930   SIMMONS SPLIT QUEEN FOUNDATION
982   930   SIMMONS KING
983   930   SIMMONS TWIN X-LONG FOUNDATION
984   930
985   930
986   930
987   930   SERTA MATTRESS
988   930   SERTA FOUNDATION
989   930   HYBRID MATTRESSES
990   990   BRASS/METAL BEDS
996   996   ADDITIONAL DELIVERY CHARGES
998   998   FABRIC PROTECTION
999   998   EXTENDED WARRANTY

64 rows selected.
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 20, 2013, 8:43:24 AM:_

Regarding calculating the bedding subtotal:

    Sum of "Bedding Only Sale" = PRODUCTCLASSID >= 930 excludes PRODUCTCLASSID in (964, 968, 949, 990)

Packages, as you know, hold their retail at the component level, not the package "header line".  Will this definition still get the correct total, or were you expecting that packages look to the package header's minor code?

**Comment by kentbower** [comment]
_**kbower** trac status change on Sep 20, 2013, 10:23:42 AM:_ **waiting**

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 20, 2013, 10:34:28 AM:_

Hi Kent,
 
It is 996M (M is in capitalized).
Itm_tp_cd = 'INV' or inventory = 'Y' you should not need both.
 
I hope that answers your questions.
 
Doug

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 20, 2013, 10:48:30 AM:_

Waiting on how to handle packages and productclass 991 and above...

**Comment by kentbower** [comment]
_**kbower** trac status change on Sep 20, 2013, 10:48:30 AM:_ **waiting**

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 20, 2013, 2:19:57 PM:_

(Changeset [[4560]](../commit/f961e6b764593e2993f034aeae505533c7b08392))
* closes #2432, Bedding Only Delivery Charges #stn
   * $50 or $89.95 discount for Bedding Only orders based on $498 bedding total threshold
   * Prompt for comment if override delivery charge
   * Same Day Mattress (SDM) Shipping can only be saved with Bedding Only orders
   * Same Day Mattress (SDM) Shipping today's delivery date deadline of 3pm
   * New permission 'stnsamedaydelivery',"ORDERS - Ability to schedule same day deliveries after 3:00pm deadline (also need 'orderdeliveryadmin')."

**Comment by kentbower** [comment]
_**kbower** trac status change on Sep 20, 2013, 2:19:57 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 23, 2013, 1:16:59 PM:_

Also changeset [[4568]](../commit/39c2642631b0f174d21126818befc23ae4a42f26)

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Jun 1, 2015, 2:31:41 PM:_

Instead of subtracting $89.95, use the current delivery charge on zone 'SAN'

```
        ('System','ShippingZone.process_charge','11','Calculate the correct charge for ShippingZone for date','Y',
'''# NOTE: deliverycharge dependencies and reset_dependencies implications!
# process as normal first because specials are based off this charge
shippingzone.process_charge(allow_rule=False)
sale = shippingzone.order
if sale.deliverytype != u'P' and \\
   sale.is_sale:
    if sale.customerorder.stn_bedding_shipping:
        if sale.customerorder.stn_bedding_shipping == u'A':
            discount = Decimal(50)
        elif sale.customerorder.stn_bedding_shipping == u'B':
            discount = Decimal('89.95')
        # if discount is undefined, allow crash
        shippingzone.charge = max(Decimal(0), (shippingzone.charge or Decimal(0)) - discount)
''',),
```

**Comment by kentbower** [comment]
_**sjohnson** trac status change on Jun 1, 2015, 2:31:41 PM:_ **reopened**

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Jun 1, 2015, 4:11:36 PM:_

(Changeset [[6456]](../commit/508ec19a91a9d2c687114ae4178c04a59c502ab2))
* closes #2432 Default bedding delivery discount to equal the charge from SAN zone #stn

**Comment by kentbower** [comment]
_**sjohnson** trac status change on Jun 1, 2015, 4:11:36 PM:_ **closed**
